src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(string fullPath, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// SourceFileInfo is random name I choose. we might want to chagne this to something like temporarySourceInfo,
+    /// InterimSourceInfo, GeneratedSourceInfo and etc.
+    /// 
+    /// none of info returned from this can be dynamic. it should always return same data
+    /// </summary>
+    internal abstract class SourceFileInfo
+    {
+        /// <summary>
+        /// unique id to identify this file among other files in a project. it should be unique in the project
+        /// </summary>
+        public abstract string Id { get; }
+        
+        /// <summary>
+        /// file path of this source file. since this file can be virtual (doesn't actually exist in the file system),
+        /// file path doesn't need to point to actual physical file in the file system.
+        /// 
+        /// this can be used with something like editorconfig to find out right options for this source file.
+        /// </summary>
+        public abstract string FilePath { get; }
+        
+        /// <summary>
+        /// return TextLoader to get content of this source file
+        /// </summary>
+        public abstract TextLoader TextLoader { get; }
```

Comment:
I think we're missing `IDocumentServiceFactory` here as well. 

Since `TextLoader` and `IDocumentServiceFactory` will change at the same time, we need to express that in a thread-safe way through the API.

Reply:
ah.. yep. I forgot to add the most important one :) let me add it.